### PR TITLE
Allow any model architecture

### PR DESCRIPTION
**Pull request by ClashLuke**

### CODE REVIEW

**Comment by ClashLuke** [comment]
Works well. example [notebook](https://colab.research.google.com/gist/ClashLuke/ee59ddb6721d54fc04739020be789401/copy-of-copy-of-locoprop_torch.ipynb?authuser=2#scrollTo=9JBN3dAqF9As)

File: locoprop/layer.py [hunk 1]
```diff
@@ -1,19 +1,106 @@
...
+
+        self.storage_for_grad = None
+        if self.lctx.correction <= 0:
             return hidden
+
+        # correct input of next layer
+        with torch.no_grad():
+            delta = self.module(*args, **kwargs) - hidden
+            correction = self.lctx.correction / delta.std().clamp(min=self.lctx.correction_eps)
+            return hidden + correction.clamp(max=1) * delta
```
Comment (dvruette) [inline locoprop/layer.py]: This is not quite the same correction behavior as before, right? Can you quickly explain why we use the inverse std of the delta instead of a constant scalar?

File: locoprop/layer.py [hunk 2]
```diff
@@ -1,19 +1,106 @@
...
+            for n, p in self.module.named_parameters():
+                p.grad = original_params[n].data - p.data
+                p.set_(original_params[n].data)
+
+        self.storage_for_grad = None
+        if self.lctx.correction <= 0:
             return hidden
+
+        # correct input of next layer
+        with torch.no_grad():
```
Comment (dvruette) [inline locoprop/layer.py:134]: We only need to execute this block if `self.lctx.correction > 0`, saving a bit of computation if correction is disabled.

File: locoprop/layer.py [hunk 3]
```diff
@@ -1,19 +1,106 @@
+import dataclasses
+from typing import Optional, Any
+
+import torch
 from torch import nn
 
+_IS_TRAINING = [False]
```
Comment (dvruette) [inline locoprop/layer.py]: What's the reason this needs to be a list? Couldn't it be just a boolean? Also, is this already compatible with MP? If not, could we easily make it so?

**Review by dvruette — COMMENTED** [review]
Looks great! Ideally we would update at least the `test/autoencoder.py` example (and the README) for future reference.

I've left some comments where I didn't quite understand.

Another thing I've noticed yesterday is that `torch.autograd.grad` doesn't work with integer tensors as inputs, which happens e.g. when using `nn.Embedding`. Not sure if this is easily resolvable, so we don't need to address it now. But something to keep in mind for the future.

File: locoprop/layer.py (hunk 1 above)
Comment (dvruette) [reply locoprop/layer.py]: For reference, it used to be:
```python
with torch.no_grad():
    delta = layer(inp) - inps[i + 1]
    norm = delta.norm() + 1e-5
    inps[i + 1] += min(norm, self.correction * delta.size(1) ** 0.5) * delta / norm
```

Comment (ClashLuke) [reply locoprop/layer.py]: `min(norm, x) / norm = min(1, x / norm) = (x / norm).clamp(max=1)`
with `x = self.correction * delta.size(1) ** 0.5` and `norm = delta.norm()`
`(x / norm).clamp(max=1) = (self.correction / delta.std()).clamp(max=1)`
for stability purposes, we can add an epsilon to the std. Here, I simply prefer clamping epsilon in place rather than adding it, so that 1e-5 won't be turned into 2e-5.

File: locoprop/layer.py (hunk 2 above)
Comment (ClashLuke) [reply locoprop/layer.py:134]: I moved the if two lines above to avoid indenting the block

File: locoprop/layer.py (hunk 3 above)
Comment (ClashLuke) [reply locoprop/layer.py]: I didn't figure out how to have one module set a boolean and the other read it. Using a list, it becomes trivial. Do you have any idea to solve this?

Comment (ClashLuke) [reply locoprop/layer.py]: It's technically not compatible with multiprocessing or model-parallel. Practically, as long as all python processes are synchronised, it should be fine. (Synchronisation is a requirement anymore for torch model-parallel)

File: locoprop/layer.py (hunk 1 above)
Comment (dvruette) [reply locoprop/layer.py]: Not sure what you mean in the first line: Suppose `norm = 2` and `x = 1`, then `min(norm, x) = 1` and `min(1, x / norm) = 0.5`.

For the second line, you're essentially saying that `delta.norm() / sqrt(dim) = delta.std()` with `dim = delta.size(1)` which I don't see where it's coming from. Suppose `delta = torch.ones(...)`, then `delta.std() = 0` and `delta.norm() / sqrt(dim) = 1`. Am I missing something?

Comment (ClashLuke) [reply locoprop/layer.py]: `min(norm, x) = 1`
but you'd still divide by norm afterwards, resulting in 0.5

you're right, the second line is wrong. it'd only be correct if it was delta.numel()

File: locoprop/layer.py (hunk 3 above)
Comment (dvruette) [reply locoprop/layer.py]: I see, should be good then. Maybe we can add a comment about this:
```suggestion
# Using a list to enable read/write operations by different modules.
# Not strictly multiprocessing-compatible, expect unexpected behavior.
_IS_TRAINING = [False]
```

File: locoprop/layer.py (hunk 2 above)
Comment (dvruette) [reply locoprop/layer.py:134]: Oh, missed that MB.

File: locoprop/layer.py (hunk 1 above)
Comment (dvruette) [reply locoprop/layer.py]: Oh right, missed the division by norm. You're absolutely right.

Second point: Even still, the result is the same for e.g. `delta = torch.ones(8)`. Here, `delta.numel() = delta.size(0)`. There might be some merit to the std-based approach, but I don't see immediately why it should work (better).

Comment (ClashLuke) [reply locoprop/layer.py]: should it also be delta.size(1) for convnets, or would `delta.numel() / delta.size(0)` be better?

Comment (ClashLuke) [reply locoprop/layer.py]: > why it should work (better).

purely code simplification. norm is likely more stable, as the epsilon is in front of the sqrt.

Comment (dvruette) [reply locoprop/layer.py]: Good point, depends on the behavior of `norm()`. Thinking about it, it probably always should have been `norm = delta.view(delta.size(0), -1).norm(1).mean()` and `(delta.numel() / delta.size(0))**0.5`. The idea is to account for growing norm magnitude with growing dimensionality. Not sure how sound this is to begin with but it seemed more stable than just the vanilla norm when I tried out both.

Comment (dvruette) [reply locoprop/layer.py]: Or probably even `delta.view(delta.size(0), -1).norm(1)` to have per-sample corrections, which makes much more sense.

Comment (ClashLuke) [reply locoprop/layer.py]: wouldn't `x.flatten(1).norm(1) * (x.numel() / x.size(0))**0.5` be equal to `x.flatten(1).std(1)`?

Comment (dvruette) [reply locoprop/layer.py]: So IIRC we have `std(x) = ((x - x.mean()).pow(2).sum() / (x.numel() - 1)).sqrt()` whereas `x.norm() = x.pow(2).sum().sqrt()`. Standard deviation quantifies the variance in the data and depends on `x`, the dimension normalizer is independent of data. (consider the case where `x = torch.ones(8)`)

Comment (dvruette) [reply locoprop/layer.py]: Tested the two approaches like this:
```python
x0, x1 = torch.ones(16, 32, 32, 3), torch.zeros(16, 32, 32, 3)
correction = 0.1

delta = x0 - x1
mag = correction / delta.flatten(1).std(1).clamp(min=1e-5).clamp(max=1)
delta = (mag.unsqueeze(-1) * delta.flatten(1)).view(*delta.shape)
res1 = x1 + delta

delta = x0 - x1
norm = delta.flatten(1).norm(dim=1).clamp(min=1e-5)
mag = (correction / norm * delta.flatten(1).size(1) ** 0.5).clamp(max=1)
delta = (mag.unsqueeze(-1) * delta.flatten(1)).view(*delta.shape)
res2 = x1 + delta

(res1 - res2).norm().sum()
>>> tensor(2217092.7500)
```

Going to reintroduce the second approach because I don't understand the logic behind the first one (it might still be justified, I just can't see it right now). We should be good to merge then.

Comment (ClashLuke) [reply locoprop/layer.py]: You're right. My assumption was that std = norm*constant, which is only correct of it's already centered.

Comment (dvruette) [reply locoprop/layer.py]: Ah I see what you meant now!

Comment (ClashLuke) [reply locoprop/layer.py]: I fused the .min() and .max() calls into one. Probably good if you have another look before merging.

**Comment by dvruette** [comment]
While training seems to work with Lightning, the standard training loop breaks for some reason: https://colab.research.google.com/drive/1hNLavl5jYgf7-DxfmnTucyCAcMHS4y4c?usp=sharing

Don't really know what's going on, the only difference is the training loop AFAICT.

**Comment by ClashLuke** [comment]
We get roughly the same convergence with an outer learning rate of 0.01. Look at [this copy](https://colab.research.google.com/gist/ClashLuke/bd5f4ff1cabdd00bfe18c6b613290544/locoprop_torch_autograd.ipynb) of your notebook.\
Does the lightning trainer do some gradient scaling internally?
